### ISSUE TITLE
Bug: [TOC] Broken

### ISSUE THREAD

**Issue by VillageHubertChen** [classeur/classeur#155]
# How the Bug Shows

![classeur-bug-toc](https://cloud.githubusercontent.com/assets/7278716/17460492/ecfd10a2-5c9b-11e6-993b-7c4b3e99f99a.png)

Preview no longer shows the rendered Table of Contents created by the [TOC] entry.
# Time of Bug Introduction

Just recently. I had a page open with working [TOC] in Chrome. A Shift + Page Reload broke it. It was completely working a few days ago.
# Reproducing
- On two different computer, using the test document below
- FireFox and Chrome
# Test Document

---

[TOC]
# Heading 1
### Heading 3

---
# If this is a feature removal
### Correct Depth of Table of Contents

You can set the depth of the Table of Contents to 2 using "Properties". We need this feature to view a depth of ToC of 2 during writing. The ToC in Side Bar is not considering "Properties" and is always showing depth of 6.

If you were to correctly implement this, then [TOC] is no longer needed.
### Table of Contents Export

We could get no Table of Contents Export to work. It is possible it works, but we could not find out how. As such our procedure is as follows:
1. Insert [TOC] in markdown → generates a Table of Content on document
2. Export document to HTML
3. Edit HTML page source to extract the HTML UL list of the TOC
4. Insert that UL HTML code into our TOC converter converting it to YML, JSON, Markdown

If you were to implement a working TOC export with format HTML UL (unordered list) we then no longer need the [TOC] function.

Let us know in case you like to include the function of exporting TOC to YAML and JSON and we send you the link to our repository. It is under MIT license.

**Comment by orschiro** [comment]
Thanks for your report, Hubert!

However, I can't reproduce it on Ubuntu Linux 16.04 using the Chrome App. 

@benweet what about you?

![image](https://cloud.githubusercontent.com/assets/170947/17460855/110d2302-5c78-11e6-9a6b-10776482e516.png)

**Comment by VillageHubertChen** [comment]
Hi Robert, @orschiro 

Thanks for the quick response! We can use it. Two important notices:
1. **Did you Shift + Reload** the page? It was working fine for me too, until I made a Shift + Reload!
2. I forgot to mention that we are using Windows 7 64 Bit, English. All latest updates installed.
3. Is there any further test you want us to do?

**Comment by orschiro** [comment]
Hubert, you're right! I overlooked the page reload. Indeed, reloading the page (I pressed F5 in the Chrome app) does break the TOC and only shows `TOC`. 

I will mark it as bug. 

Thanks!

**Comment by VillageHubertChen** [comment]
Hi Robert, @orschiro

Thanks for retesting. Glad you could reproduce it :-)

**Comment by VillageHubertChen** [comment]
# How to export ToC until Bug is fixed

For anyone who need to export a Table of Contents now:
1. Add [TOC] to the top of your markdown document
2. Copy your markdown document into the clipboard
3. Paste it into → https://stackedit.io/editor#
4. Press the "Open in Viewer" Button
5. Right Click "Inspect Element"
6. Click the "Pick an Element of the Page" button (most left button, arrow in rectangle)
7. Click on the generated Table of Contents
8. Right click in the source code on `<div class="toc">`
9. Select in drop down menu: Copy Inner HTML
10. Paste the HTML UL TOC into the [TOC Converter](https://v-squared.github.io/v2-ToC-Converter/)
11. Copy Paste the JSON or YAML version of your TOC depending on your Publication System requirements

P.S. We will add a Markdown version of the TOC soon, so you can republish a partial ToC. Say you need a partial ToC as introduction to a lengthy chapter within the article.

**Comment by VillageHubertChen** [comment]
Hi Ben & Robert,

It has been 2 weeks. What is the status of fixing the ToC Bug? Are you waiting on something from me?

**Comment by orschiro** [comment]
Dear @benweet,

Can you give @VillageHubertChen an estimation?

Thankfully,

~Robert

**Comment by benweet** [comment]
Hi @VillageHubertChen 

Indeed the `[TOC]` marker was deprecated (see https://github.com/classeur/classeur/issues/129#issuecomment-228868271). It was not standard, not supported by Pandoc and redundant with the Table of Content feature provided by Pandoc.
However, you do have a point, table of content should be available in the HTML export.

Here is my proposition:
- a `toc` structure will be exposed as follow in the Handlebars context (see https://classeur.zendesk.com/hc/en-us/articles/208710309):

``` json
{
  "file": {
    "name": "The filename",
    "content": {
      "properties": {},
      "text": "...",
      "html": "...",
      "toc": [
        {
          "title": "Heading level 1",
          "children": [
            {
              "title": "Heading level 2",
            }
          ]
        },
        {
          "title": "Another heading level 1",
        }
      ]
    }
  }
}
```
- A `tocToHtml` template helper will be used to convert the toc object into HTML format:

```
{{#tocToHtml file.content.toc}}{{/tocToHtml}}
```

An optional parameter can be used to limit the depth of the Table Of Contents (for example level 2):

```
{{#tocToHtml file.content.toc 2}}{{/tocToHtml}}
```

As for the Table Of Contents in the sidebar, would it be enough to have a global `depth` parameter in the settings?

Cheers

**Comment by VillageHubertChen** [comment]
Dear @benweet 

Thanks for your swift response. @orschiro and me are working on a reply. We will post it as soon we have it.

**Comment by VillageHubertChen** [comment]
Dear @benweet,

After discussion with Robert we propose the following:

Add 4 Export TOC buttons in the ToC Side-Bar:
1. to HTML
2. to JSON
3. to Markdown
4. to YAML

After pressing the button the ToC content is copied to the clip-board. The author can use it in many ways.

The three other format conversions can be done using: https://v-squared.github.io/v2-ToC-Converter/

You can use our source code.

More details via Robert's phone call. Let me know when you have any questions.

Good luck,

~Hubert

**Comment by orschiro** [comment]
Dear @VillageHubertChen and @benweet,

I want to summarise what we had agreed upon.

We will provide a solution for people to export the ToC by the following workflow:
1. Open a document
2. Open the document menu
3. Click on Export
4. Select the template "Export ToC" from the dropdown list

**Further remarks:**

The template "Export ToC" will become part of a future template repository that users can contribute to and share reusable templates. 

Is everyone of us satisfied now? :-)

Warmly,

-Robert

![image](https://cloud.githubusercontent.com/assets/170947/18316006/5372cf80-7519-11e6-8863-f4b0087a7363.png)

**Comment by VillageHubertChen** [comment]
Dear @benweet and @orschiro ,

Thank you so much for your efforts. These are my questions to clarify the details:
1. The Export ToC would be in HTML UL format, just like the example on the bottom?
2. There will be a drop down list, where the user can choose the depth of the ToC, ranging from 1 ... 6? 
   - Having the depth of ToC as a generic settings somewhere in preferences would not do, as some users would need to change the frequently
3. The export is directly to clipboard? If instead you export to a file, can you add an option to export to clipboard? That would make a big difference in the workflow, where you need to paste the HTML ToC into the HTML ToC Converter and then get the usable ToC format in Markdown or YAML.
4. This would be the dream: Do you provide three templates for ToC Export?
   - HTML ToC
   - Markdown ToC
   - YAML ToC

Warmly,

~Hubert

## Addendum: Example Code of HTML ToC

``` HTML
<ul>
   <li>
      <a href="#table-of-contents">Adding Table of Contents</a>
      <ul>
         <li><a href="#manual-vs-automatic">Manual versus Automatic ToC</a></li>
         <li><a href="#manual-toc">Manual ToC</a></li>
         <li><a href="#automatic-toc">Automatic ToC</a></li>
      </ul>
   </li>
   <li>
      <a href="#advanced-table-formatting">Advanced Table Formatting</a>
      <ul>
         <li><a href="#overview-advanced-tables">Overview</a></li>
         <li><a href="#polish-code">Polish Table Source Code</a></li>
         <li><a href="#apply-styling">Apply Styling</a></li>
         <li><a href="#use-graphic-symbols">Use Graphic Symbols</a></li>
      </ul>
   </li>
   <li>
      <a href="#super-table-formatting">Super Table Formatting</a>
      <ul>
         <li><a href="#user-story">User Story</a></li>
         <li><a href="#super-tables-features">Key Features</a></li>
         <li><a href="#how-it-works">How it works</a></li>
         <li><a href="#images-in-tables">Images in Tables</a></li>
         <li><a href="#workflow-tables">Workflow for Complex Tables</a></li>
         <li><a href="#ref-background-colors">Reference of all Background Colors</a></li>
         <li><a href="#ref-table-styles">Reference for all Table Styles</a></li>
      </ul>
   </li>
   <li>
      <a href="#slide">SlideShow</a>
      <ul>
         <li><a href="#slide-example">Example</a></li>
      </ul>
   </li>
   <li>
      <a href="#reference">Reference</a>
      <ul>
         <li><a href="#ref-tools">Tools</a></li>
         <li><a href="#graphic-symbols">Symbols used in Tables</a></li>
      </ul>
   </li>
</ul>
```

**Comment by orschiro** [comment]
Thanks for your remarks, @VillageHubertChen!

@benweet can I leave this up to you?

-Robert

**Comment by benweet** [comment]
Hi @VillageHubertChen 
1. The TOC template would look like this:
   
   ```
   {{#tocToHtml file.content.toc}}{{/tocToHtml}}
   ```
   
   And would produce the HTML you described.
2. To have a limited depth (let's say 3), you would have to modify or create a template with this:
   
   ```
   {{#tocToHtml file.content.toc 3}}{{/tocToHtml}}
   ```
3. It would export to a file. Export to clipboard would be a separate feature (ideally another ticket should be created for that)
4. Producing YAML, JSON or Markdown instead of HTML is a matter of creating a custom Handlebars helper (we could assist you for that).

**Comment by VillageHubertChen** [comment]
Hi @benweet , thanks. See my answers below:
Hi @orschiro , thanks. Can you take care of item 3. ?
1. Understood. Great! (Closed)
2. So I could create 6 templates to provide the 6 depths? (If you confirm than this is closed)
3. The ability to export directly to the clipboard will make a difference. I suppose if Robert opens the ticket will be more convenient for you as he knows your style what information you would want to see in it. (closed if Robert opens issue)
4. Fascinating. Sound powerful. I am looking forward to that. (closed)

Thanks so much to both of you!

~Hubert

**Comment by orschiro** [comment]
Dear @VillageHubertChen and @benweet,

Point 3, the export to clipboard, can be further discussed here: https://github.com/classeur/classeur/issues/188

-Robert

**Comment by VillageHubertChen** [comment]
Dear @orschiro 

Perfect! Thank you!

~Hubert
